luavm/src/lib/lua_values/mod.rs

```diff
@@ -0,0 +1,754 @@
+mod lua_table;
+mod tagging;
+
+use self::{lua_table::LuaTable, tagging::*};
+use errors::LuaError;
+use gc::{Finalize, Gc, Trace};
+
+/// Represents a value in Lua.
+#[derive(Debug)]
+pub struct LuaVal {
+    encoded_ptr: usize,
+    /// whether this should drop the pointer or not
+    owns: bool,
```

Comment:
I don't understand exactly what `owns` is doing: it feels like it's a sort-of GC-ish hack based on shallow copying, but it's not clear to me that it's safe or needed. Can you explain in more detail?

Reply:
I had to come up with this 'hack' because I can't seem to be able to return a reference from `get_attr` in `LuaTable`. At the moment `get_attr` returns a shallow copy of the attribute, so that if the interpreter modifies it, the dict's state is updated correctly.

Reply:
Shouldn't we be getting this sort of behaviour with the `Gc` stuff? I'm not even sure this hack is safe TBH.

Reply:
Now that I think about it, should `get_attr` return a reference to the value? If a user needs to set an attribute, they would use `set_attr`, not `get_attr` followed by a set.

Reply:
If you return a Rust reference, I doubt very much that you can model Lua (or even compile the resulting Rust program!).

What I thought should be the case is that `LuaVal.clone` (which you might call `LuaVal.shallow_clone` but it doesn't matter) simply copies the LuaVal (if it's an int etc.) or clones the Gc object (if it's a Gc'able thing).